packages/react-native/Libraries/NativeAnimation/RCTAnimationUtils.mm

```diff
@@ -94,12 +95,20 @@ uint32_t RCTInterpolateColorInRange(CGFloat value, NSArray<NSNumber *> *inputRan
   [outputRange[rangeIndex] getRed:&redMin green:&greenMin blue:&blueMin alpha:&alphaMin];
   CGFloat redMax, greenMax, blueMax, alphaMax;
   [outputRange[rangeIndex + 1] getRed:&redMax green:&greenMax blue:&blueMax alpha:&alphaMax];
-
-  return RCTColorFromComponents(
-      0xFF * (redMin + (value - inputMin) * (redMax - redMin) / (inputMax - inputMin)),
-      0xFF * (greenMin + (value - inputMin) * (greenMax - greenMin) / (inputMax - inputMin)),
-      0xFF * (blueMin + (value - inputMin) * (blueMax - blueMin) / (inputMax - inputMin)),
-      alphaMin + (value - inputMin) * (alphaMax - alphaMin) / (inputMax - inputMin));
+    
+  CGFloat r = redMin + (value - inputMin) * (redMax - redMin) / (inputMax - inputMin);
+  CGFloat g = greenMin + (value - inputMin) * (greenMax - greenMin) / (inputMax - inputMin);
+  CGFloat b = blueMin + (value - inputMin) * (blueMax - blueMin) / (inputMax - inputMin);
+  CGFloat a = alphaMin + (value - inputMin) * (alphaMax - alphaMin) / (inputMax - inputMin);
+    
+//  return [UIColor colorWithRed:r green:g blue:b alpha:a];
+  return @{
+    @"space": @"srgb",
```

Comment:
@cipolleschi I'm still struggling with how to handle the unclamped values here. If I understand Apple's docs correctly we'd need to pass those to colorWithRed:green:blue:alpha: so we'd want to stay in extended srgb color space. But this doesn't appear to work. 

Also I'm not sure I understand why I can't just use UIColor. When I try to return the UIColor directly here or from RCTColorAnimatedNode I don't get the color to come through at all.

Reply:
I'm not sure I completely understand your questions, but:

> If I understand Apple's docs correctly we'd need to pass those to colorWithRed:green:blue:alpha: so we'd want to stay in extended srgb color space. But this doesn't appear to work.

This is true if the source space is sRGB.
If we are in displayP3, values that are < 0 and > 1 will not be properly represented. 🤔 

I think we need to create the final `UIColor` by using `colorWithDisplayP3Red:green:blue:alpha:`. 

> Also I'm not sure I understand why I can't just use UIColor. When I try to return the UIColor directly here or from RCTColorAnimatedNode I don't get the color to come through at all

If we change the return type from `int32_t` to something else, we have to update the callsites to understand the new return type, no?

I'm not super familiar with this part of the codebase, I need to dig deeper here.
---
If this is the only blocker, I think we can move forward for the time being and consider this a known-issue: **for the first implementation interpolateColor don't work with displayP3**.

Even with this limitation, this would already be a massive improvement! I think that animating colors is not something every app actually require, so it is an edge case that it is ok to point out as "not working" for version 1 of the implementation.

Reply:
Using `colorWithDisplayP3Red:green:blue:alpha:` gives me the same behavior as before including the color space and I believe it's because [per Apple](https://developer.apple.com/documentation/uikit/uicolor/1648569-colorwithdisplayp3red?language=objc):
> The red component of the color object, specified as a value from 0.0 to 1.0.

and

> Values below 0.0 are interpreted as 0.0, and values above 1.0 are interpreted as 1.0.

Meanwhile with [`colorWithRed:green:blue:alpha:`](https://developer.apple.com/documentation/uikit/uicolor/1621930-colorwithred?language=objc) 
> The red value of the color object. On applications linked for iOS 10 or later, the color is specified in an extended color space, and the input value is never clamped. On earlier versions of iOS, red values below 0.0 are interpreted as 0.0, and values above 1.0 are interpreted as 1.0.

So I believe if we have unclamped values we may actually need to use `colorWithRed:green:blue:alpha:` to have them interpreted correctly. But then again I couldn't get it to work completely either way.

As far as returning UIColor directly goes I updated the call sites for [`RCTInterpolateColorInRange`](https://github.com/facebook/react-native/blob/782e9eace974baefba7d8978739f11fe1ea70dc9/packages/react-native/Libraries/NativeAnimation/Nodes/RCTInterpolationAnimatedNode.mm#L153C24-L153C50) and [`RCTColorAnimatedNode`](https://github.com/facebook/react-native/blob/main/packages/react-native/Libraries/NativeAnimation/Nodes/RCTStyleAnimatedNode.mm#L53) but I was having trouble following what happens to the values after `performUpdate` is called. Essentially I'd like to know how the animated values ultimately get applied back to the original views.

 <hr>
 
> If this is the only blocker, I think we can move forward for the time being and consider this a known-issue: for the first implementation interpolateColor don't work with displayP3.
>
> Even with this limitation, this would already be a massive improvement! I think that animating colors is not something every app actually require, so it is an edge case that it is ok to point out as "not working" for version 1 of the implementation.

Makes sense. I'll move on to Android today.

Reply:
To explore how the colors change based on the rgb values, I created this small project: https://github.com/cipolleschi/P3Colors.

feel free to clone it and play with it.

The top-left view is sRGB
The to-right view is displayP3
The bottom-left view is sRGB when we passed in the rgb values obtained by `getRed:green:blue:alpha` method on the displayP3 color.
The bottom-right view is displayP3 when we passed in the rgb values obtained by `getRed:green:blue:alpha` method on the displayP3 color.

Use a retina display to look at it! 😄 

https://github.com/ryanlntn/react-native/assets/11162307/122550c5-18b1-4c8a-b77b-623481ba0f8f

Reply:
![IMG_0905](https://github.com/ryanlntn/react-native/assets/1944151/4120a97c-5025-431e-a963-950f9caa9a07)

Interesting. It looks like my interpretation of Apple's docs is correct then since passing the DisplayP3 colors to `colorWithRed:green:blue:alpha:` gives the same color as the original DisplayP3 color while passing to `colorWithDisplayP3Red:green:blue:alpha:` is too red up until you hit values greater than 1 at which point it's the same.

The behavior right now in `RCTInterpolateColorInRange` is even more bizarre then since it's currently cycling through yellow and blue as well. 🤔

Reply:
yeah... it hink that the behavior of interpolate is wrong in the context of display P3 because the color space is non euclidean. So the usual interpolation formula (which are linear) would not work properly there... 😓 
Unfortunately, I don't have a clear solution... 
One idea we can try to:
- preprocess: interpolate the starting color from displayP3 to sRGB representation (e.g. `red` in P3 to `red` in sRGB)
- interpolation: interpolate the colors in sRGB, from the source to the destination (e.g from red to blue) reusing the sRGB formula
- postprocess: interpolate the final color from sRGB to displayP3 (e.g. `blue` in P3 to `blue` in sRGB).

In that way we minimize the weirdness paying the tradeoff that interpolation is in sRGB... but if that happens mainly in animation, they should be "fast" enough so that the user won't notice that they are not in displayP3 space...

What do you think?